corehq/apps/case_search/utils.py

```diff
@@ -47,14 +48,28 @@
 from corehq.util.timer import TimingContext
 
 
+@dataclass
+class CaseSearchProfiler:
+    debug_mode: bool = False
+    primary_count: int = 0
+    related_count: int = 0
+    timing_context: TimingContext = field(
+        default_factory=lambda: TimingContext('Case Search'))
+    queries: list = field(default_factory=list)
+
+    def add_query(self, slug, es_query):
```

Comment:
I thought about making this instead `run_query`, so I could do both the timing context and query logging inside here, but I don't love moving core logic like that in here.  Dunno, open to suggestions.  It'd look something like:
```python
results = profiler.run(query, "subcase_query")
```
instead of
```python
profiler.add_query("subcase_query", query)
with profiler.timing_context("subcase_query"):
    results = query.run()
```